doc: add CONTRIBUTING.rst to project repo and docs

dbkinder opened the pull request:
GitHub notices a CONTRIBUTING file in a repo's root and will automatically
add a link to this file on the page when a contributor creates an Issue
or opens a Pull Request. (Expectation is CONTRIBUTING will have
information about how to contribute to the project, format code,
test fixes, and submit patches.

We also want to have this document accessible from our technical docs,
and not duplicate the content, so add linkage to make this work.

The zephyrproject github wiki article that contributed to this new
CONTRIBUTING doc will be made into a referece to this new doc once
this PR is approved and merged.

Replaces PR #929

Jira: ZEP-2085

Signed-off-by: David B. Kinder <david.b.kinder@intel.com>

dbkinder committed:
doc: add CONTRIBUTING.rst to project repo and docs

GitHub notices a CONTRIBUTING file in a repo's root and will
automatically
add a link to this file on the page when a contributor creates an Issue
or opens a Pull Request. (Expectation is CONTRIBUTING will have
information about how to contribute to the project, format code,
test fixes, and submit patches.

We also want to have this document accessible from our technical docs,
and not duplicate the content, so add linkage to make this work.

The zephyrproject github wiki article that contributed to this new
CONTRIBUTING doc will be made into a reference to this new doc once
this PR is approved and merged.

Replaces PR #929

Jira: ZEP-2085

Signed-off-by: David B. Kinder <david.b.kinder@intel.com>

dbkinder commented:
Please see if you agree with the workflow path I based on Anas' wiki article and amended.  I didn't want to get into how to break up commits (that's a separate problem), but is using ``commit --amend`` and force-pushing an acceptable workflow for making reviewer updates?  (That how I've been working on docs.)  Feel free to making specific workflow recommendations (as I'm sure you will :) with the workflow you'd prefer.   P.S. PR #929 can be closed and this PR superseeds that one.

inakypg commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+*********
+
+Licensing is very important to open source projects. It helps ensure the
+software continues to be available under the terms that the author desired.
```
s/authors agreed to/?

inakypg reviewed (commented):
I like it but I expect most people will TL;DR

Here is what I would suggest: a super compressed version with the "directions"that takes no longer than a page and has embedded to the longer sections that include the rationales.

nashif commented:
> but is using commit --amend and force-pushing an acceptable workflow for making reviewer updates? (That how I've been working on docs.) 

You have a better option?

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+
+    git clone https://github.com/zephyrproject-rtos/zephyr
+
+(You may find `using ssh access to GitHub`_ is more convenient since you
```
Cloning the mainline repository shouldn't require a username or password, since it's public.

Also note that you can use a credential helper to cut down on having to enter your password, even with https: https://git-scm.com/docs/gitcredentials

dbkinder replied:
removed

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+.. _using SSH access to GitHub:
+   https://help.github.com/articles/connecting-to-github-with-ssh/
+
+The Zephyr project directory structure is describe in `Source Tree Structure`_
```
described

dbkinder replied:
done

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+
+.. _Zephyr Project Issues: https://jira.zephyrproject.org
+
+Before starting on a patch, first check in our Jira `Zephyr Project Issues`_
```
JIRA? If so, to be applied below as well.

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+
+Before starting on a patch, first check in our Jira `Zephyr Project Issues`_
+system to see what’s been reported on the issue you’d like to address.  Have a
+conversation on the mailing list (or IRC) to see what others think of your
```
links may make this easier:

- https://lists.zephyrproject.org/mailman/listinfo
- irc://irc.freenode.net/zephyrproject

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+system to see what’s been reported on the issue you’d like to address.  Have a
+conversation on the mailing list (or IRC) to see what others think of your
+issue (and proposed solution).  You may find others that have encountered the
+issue you're finding, or with similar ideas for changes or additions.  Send a
```
s/or with/or that have/

dbkinder replied:
done

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+Install gitlint and run it locally in your tree and branch where your patches
+have been committed::
+
+     $ sudo pip install gitlint
```
This appears to work when installed with pip3 as well, so it's probably best to document that here for consistent use of Python 3 tools throughout the project.

dbkinder replied:
done

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+have been committed::
+
+     $ sudo pip install gitlint
+     $ gitlint
```
This only checks HEAD (the most recent commit). You might want to adjust the text to recommend running it after each commit, or using the `--commits` option to specify a commit range covering all the development branches to be submitted.

dbkinder replied:
added

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+Coding Style
+************
+
+Use this coding guideline to ensure that your development complies with the
```
these coding guidelines

dbkinder replied:
done

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+* Add braces to every ``if`` and ``else`` body, even for single-line code
+  blocks. Use the ``--ignore BRACES`` flag to make *checkpatch* stop
+  complaining.
+* Use hard tab stops. Set the tab width 8 spaces.
```
to 8 spaces

dbkinder replied:
done.

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+     $ git checkout master
+     $ git checkout -b fix_comment_typo
+
+   Some Zephyr subsystems do development work on a separate branch from
```
It would be good to have a list of these subsystems, and their policies, definitively documented somewhere. I've observed this can be a stumbling block for beginners.

dbkinder replied:
Indeed... worth adding later.

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+
+   or if working on a different branch than ``master`` (as mentioned above)::
+
+     $ git push origin/net fix_out_of_date_patch
```
This should be deleted. "git push origin $BRANCH" is all that's needed.

dbkinder replied:
ok

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+
+#. Review the pull request changes, and verify that you are opening a
+   pull request for the appropriate branch. The title and message from your
+   commit message should appear as well.
```
This workflow assumes that the pull request contains exactly one commit, which isn't the common practice and goes against the principle of making a series of small changes that are easier to rebase etc.

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+   here, for example, by changing the base branch from ``master`` to ``net``.
+
+#. GitHub will assign one or more suggested reviewers (based on the
+   CODEOWNERS file in the repo) and you can also select additional reviewers
```
General members of the public can't select reviewers on a pull request; this is a privileged operation restricted to authorized members of the organization. (Even though I'm a "Collaborator" on this project, I can't add reviewers to my PRs, e.g.).

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+   commit message should appear as well.
+
+#. If you're working on a subsystem branch that's not ``master``,
+   you may need to change the intended branch for the pull-request
```
s/pull-request/pull request/, and below as well.

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+
+#. Click on the submit button and your pull-request is sent and awaits
+   review.  Email will be sent as review comments are made, or you can check
+   on your pull-request at https://github.com/zephyrproject-rtos/zephyr/pulls.
```
https://github.com/pulls may be better -- that's a personalized list of your account's open pull requests. Cuts out all the other open Zephyr PRs.

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+   and use the same process described above to work on this new topic branch.
+
+#. If reviewers do request changes to your patch, go back to your local repo,
+   verify you're on the right branch, make changes, ammend your commit, rebase
```
s/ammend/amend/

Here again the workflow assumes the PR has only one commit. Suggest folding the recommendations to use interactive rebase into  the main workflow sequence instead.

The `--autosquash` option to `git rebase -i` is particularly useful for handling review feedback of the form "you forgot to add a space here"; maybe worth a mention.

nashif replied:
ok, not sure where this came from, it is not in the original contribution guide in the wiki, https://github.com/zephyrproject-rtos/zephyr/wiki/Contribution-Guide

And this duplicates (and conflicts with?) the section below about rebasing "Rebasing commits to fix code review issues".

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+* A short and descriptive subject line that is less than 72 characters,
+  followed by a blank line. The subject line must include a prefix that
+  identifies the subsystem being changed, folowed by a colon, and a short
+  title, for example:  ``doc: update wiki references to new site``
```
Might be worth recommending "git log $FILE_YOU_TOUCHED" to get examples of what other people have used as a prefix, for those in doubt.

dbkinder replied:
good idea, done.

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+
+All changes and topics sent to GitHub must be well-formed, as described above.
+
+Comment Message Body
```
s/Comment/Commit/

dbkinder replied:
doh, done.

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+Other Commit Expectations
+=========================
+
+* Commits must build cleanly when applied in top of each other, thus avoiding
```
s/in top/on top/

dbkinder replied:
done

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+  separate commits and pull-requests.
+
+* When a change is included in the set to enable other changes, but it will
+  not be part of the final set, let the reviewers know this.
```
What does this mean?

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+* When a change is included in the set to enable other changes, but it will
+  not be part of the final set, let the reviewers know this.
+
+* You can pull-request RFCs (requests for comments) to send work proposals,
```
Perhaps:

You can add the "RFC" (request for comments) prefix to your pull request title to send work proposals,

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+  not be part of the final set, let the reviewers know this.
+
+* You can pull-request RFCs (requests for comments) to send work proposals,
+  progress snapshots of your work, or to get early feedback on features or
```
s/progress/in-progress/

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+  changes that will affect multiple areas in the code base.
+
+* If you are submitting a change that has someone else's Signed-off-by, please
+  be sure that you explicitly include that person as a Reviewer in Github.
```
Not possible for the reasons described above.

dbkinder replied:
removed.

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+When adding a new file to the tree, it is important to detail the source of
+origin on the file, provide attributions, and detail the intended usage. In
+cases where the file is an original to Zephyr, the commit message should
+include the following ("Original" is the assumption of no Origin tag is
```
s/of no/if no/

dbkinder replied:
done

mbolivar commented:
>> but is using commit --amend and force-pushing an acceptable workflow for making reviewer updates? (That how I've been working on docs.)

> You have a better option?

Emphasizing interactive rebase seems important for getting people to split their series up into multiple patches, no?

nashif commented:
> Emphasizing interactive rebase seems important for getting people to split their series up into multiple patches, no?

the --amend method assume interactive rebase, that is what documented AFAIK.

mbolivar commented:
> the --amend method assume interactive rebase, that is what documented AFAIK.

The current patch talks about adding a single commit, then seeing that commit title and message appear as the pull request title and description. This means the workflow assumes one commit in the PR. It doesn't get to talking about interactive rebase until after the workflow description is over, in fact. So I disagree.

nashif commented:
> The current patch talks about adding a single commit, then seeing that commit title and message appear as the pull request title and description. This means the workflow assumes one commit in the PR. It doesn't get to talking about interactive rebase until after the workflow description is over, in fact. So I disagree.

Ok, I see now, this all was not in the original contribution guidelines and it seems we are describing two methods for addressing feedback. My assumption was this would take the existing guidelines as is and just add the DCO. 

For the sake of getting the DCO integrated ASAP, I would recommend to take the existing guidelines as is, add the DCO and if we have any edits or enhancements, do them in another commit.

dbkinder commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,500 @@
+   https://www.zephyrproject.org/doc/getting_started/getting_started.html
+
+As a Linux developer, you should also be familiar with common development
+tools such as Git, GitHub, and make. The preferred development environment is
```
done

dbkinder commented:
OK, I made the suggested edits, and removed my "single commit per PR" method (that's the only one I've used for doc edits) replacing it with the interactive rebase method Anas had (that I had moved into a separate section).  

I do agree with Inaky that it would be better to have a very brief CONTRIBUTING file that referenced the larger text shown here. You all OK merging this in now I'll work on a summary version before we make the 1.9 release?

I also notice we never talk about keeping your personal fork updated with the project master (origin and upstream).  How and when do you folks recommend doing that?  I do this occasionally:

     git checkout master
     git fetch upstream
     git merge upstream/master
     git push origin master

nashif commented:
IMO all needs to be in one document, summaries for this type of document will mean that people will read the summary and skip the rest and make assumptions. This is a set of rules and we really need people to see them in one place.  So the way it now, it is perfect IMO and should be kept as such. You can add a paragraph at the top explaining what that is and probably have a table of content, but splitting this will make it less useful.

dbkinder commented:
OK.  In that case, any other changes for now?

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,480 @@
+later in this document), the developer simply adds a ``Signed-off-by``
+statement and thereby agrees to the DCO.
+
+When a developer submits a patch, is a committment that the contributor has
```
"it is a commitment"

dbkinder replied:
done

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,480 @@
+
+Before starting on a patch, first check in our Jira `Zephyr Project Issues`_
+system to see what’s been reported on the issue you’d like to address.  Have a
+conversation on the mailing list (or IRC) to see what others think of your
```
I still think it'd be better to have links to the mailing list and IRC here, but I'm OK merging this without them.

dbkinder replied:
added

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,480 @@
+=========
+
+When you submit a pull request to the project, a series of checks are
+performed to verify your commit messages meet the requirements, the same step
```
run-on sentence -- period before "the same step" ?

dbkinder replied:
done

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,480 @@
+     $ gitlint
+
+Note, gitlint only checks HEAD (the most recent commit), so you should run it
+after each commit, or use the ``--commits`` option to spefify a commit range
```
specify

dbkinder replied:
done

nashif commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,480 @@
+
+Linux is the preferred host environment for Zephyr development. As a Linux
+developer, you should be familiar with common tools such as Git and Make, and
+platforms such as GitHub.
```
I would remove this, lets not make preferences, our goal is to work well on all OSes

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,480 @@
+* Add braces to every ``if`` and ``else`` body, even for single-line code
+  blocks. Use the ``--ignore BRACES`` flag to make *checkpatch* stop
+  complaining.
+* Use hard tab stops. Set the tab width to 8 spaces.
```
Actually, come to think of it, hard tab stops with 8 space tabs is not an exception to the Linux style; it's the same on Linux:

https://www.kernel.org/doc/html/v4.12/process/coding-style.html#indentation

This line should be removed.

dbkinder replied:
ok

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,480 @@
+  blocks. Use the ``--ignore BRACES`` flag to make *checkpatch* stop
+  complaining.
+* Use hard tab stops. Set the tab width to 8 spaces.
+* Break lines at 80 characters.
```
Ditto, Linux also (in principle) obeys the 80 column rule: https://www.kernel.org/doc/html/v4.12/process/coding-style.html#breaking-long-lines-and-strings

dbkinder replied:
ok

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,480 @@
+   here, for example, by changing the base branch from ``master`` to ``net``.
+
+#. GitHub will assign one or more suggested reviewers (based on the
+   CODEOWNERS file in the repo) and you can also select additional reviewers
```
Still not true -- people can't select additional reviewers themselves without special privileges.

mbolivar replied:
Is this going to be left as-is? People will be confused about why they can't add reviewers.

nashif replied:
probably add that you have to be a project member to do that...

dbkinder replied:
done

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,480 @@
+   commit(s) to fix review issues.  In your development repo::
+
+     $ git fetch --all
+     $ git rebase --ignore-whitespace origin master
```
upstream/master, not "origin master", to avoid having to keep the forked master in sync

dbkinder replied:
ok, but origin master was in Anas wiki page

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,480 @@
+  separate commits and pull requests.
+
+* When a change is included in the set to enable other changes, but it will
+  not be part of the final set, let the reviewers know this.
```
I still don't know what this means. All patches in a series are going to get merged -- unless you're discussing adding a patch then reverting it in the same series, which wouldn't make sense. What am I missing?

nashif replied:
agree, I am not sure what this means.

dbkinder replied:
this is a cut/paste from Anas wiki article :)

nashif replied:
still, i copied that from somewhere anyways, does not apply to how we do things.

mbolivar reviewed (changes requested):
A few more things; mostly just typos that got added in changes to v2.

nashif commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,480 @@
+
+* Each commit must address a single identifiable issue and must be
+  logically self-contained. Unrelated changes should be submitted as
+  separate commits and pull requests.
```
remove "pull-requests" here. We do not want to have multiple pull-requests when we are addressing a feature with multiple commits that might be unrelated but dependent.

dbkinder replied:
ok

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,482 @@
+later in this document), the developer simply adds a ``Signed-off-by``
+statement and thereby agrees to the DCO.
+
+When a developer submits a patch, it is a committment that the contributor has
```
"commitment" is still misspelled.

dbkinder replied:
done

mbolivar commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,482 @@
+* Add braces to every ``if`` and ``else`` body, even for single-line code
+  blocks. Use the ``--ignore BRACES`` flag to make *checkpatch* stop
+  complaining.
+* Use hard tab stops.
```
The entire line should be removed -- Linux uses hard tab stops too.

dbkinder replied:
ok

mbolivar reviewed (changes requested):
Almost there, from my perspective at least.

Funny about the copy/paste stuff :)

dbkinder commented:
Fixed some UTF-8 characters (smart quotes) that snuck in, and removed the "* When a change is included in the set to enable other changes, but it will not be part of the final set, let the reviewers know this."  since neither of you knew what this meant.  That should do it for now.
